File: Data Structures & Algorithms/count-paths/submission-2.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        cache = [[-1] * n for _ in range(m)]

        def dfs(r, c):
            if r == (m - 1) and c == (n - 1):
                return 1
            elif r >= m or c >= n:
                return 0
            if cache[r][c] == -1:            
                cache[r][c] = dfs(r + 1, c) + dfs(r, c + 1)
            return cache[r][c]
        
        return dfs(0, 0)
        

File: Data Structures & Algorithms/house-robber/submission-2.cpp
class Solution {
public:
    vector<int> cache;

    int rob(vector<int>& nums) {
        cache.resize(nums.size(), -1);
        return max(findPath(nums, 0), findPath(nums, 1));
    }

    int findPath(vector<int>& nums, int curr) {
        cout << curr << endl;
        if (curr >= nums.size()) {
            return 0;
        }
        int maxMoney = 0;
        for (int i = curr + 2; i < nums.size(); i++) {
            if (cache[i] == -1) {
                cache[i] = findPath(nums, i);
            }
            maxMoney = max(maxMoney, cache[i]);
        }
        return nums[curr] + maxMoney;
    }
};
